<h1>Importing Libraries</h1>

In [ ]:
import numpy as np
import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print("Libraries imported.")

<h1>Scrapping web data</h1>

In [3]:
data = requests.get("https://en.wikipedia.org/wiki/Category:Suburbs_of_Mumbai").text
soup = BeautifulSoup(data, 'html.parser')

In [4]:
neighborhoodList = []

In [5]:
for row in soup.find_all("div", class_="mw-category")[0].findAll("li"):
    neighborhoodList.append(row.text)

In [6]:
Ml_df = pd.DataFrame({"Neighborhood": neighborhoodList})

Ml_df.head()

Neighborhood
0          Andheri
1  Anushakti Nagar
2       Baiganwadi
3           Bandra
4          Bhandup

In [7]:
Ml_df.shape

(42, 1)

<h1>Get the geographical coordinates</h1>

In [8]:
!pip install geocoder
import geocoder
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Mumbai, India'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

     |████████████████████████████████| 102kB 9.5MB/s ta 0:00:011


In [11]:
coords = [ get_latlng(neighborhood) for neighborhood in Ml_df["Neighborhood"].tolist() ]

In [12]:
coords

[[19.118459378296492, 72.84176321065843],
 [19.042830000000038, 72.92734000000007],
 [19.062940000000026, 72.92663000000005],
 [19.054370000000063, 72.84017000000006],
 [19.145560000000046, 72.94856000000004],
 [19.229360000000042, 72.85751000000005],
 [19.208660000000066, 72.82612000000006],
 [19.06218000000007, 72.90241000000003],
 [19.250030000000038, 72.85907000000003],
 [19.224690000000066, 72.86605000000003],
 [19.212750000000028, 73.08324000000005],
 [19.00534722389655, 72.85580272012932],
 [19.08652321008152, 72.90900774216628],
 [19.164550000000077, 72.84946000000008],
 [18.959290000000067, 72.83108000000004],
 [19.137920000000065, 72.84941000000003],
 [19.014920000000075, 72.84522000000004],
 [18.953937419095155, 72.82036732944775],
 [19.21198153260436, 72.83757275783374],
 [19.131380000000036, 72.93568000000005],
 [19.127580000000023, 72.82539000000008],
 [19.064980000000048, 72.88069000000007],
 [19.21198153260436, 72.83757275783374],
 [19.048530000000028, 72.93222000000003

In [13]:
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [14]:
Ml_df['Latitude'] = df_coords['Latitude']
Ml_df['Longitude'] = df_coords['Longitude']

In [15]:
print(Ml_df.shape)
Ml_df

(42, 3)


Neighborhood   Latitude  Longitude
0                     Andheri  19.118459  72.841763
1             Anushakti Nagar  19.042830  72.927340
2                  Baiganwadi  19.062940  72.926630
3                      Bandra  19.054370  72.840170
4                     Bhandup  19.145560  72.948560
5                    Borivali  19.229360  72.857510
6                     Charkop  19.208660  72.826120
7                     Chembur  19.062180  72.902410
8                     Dahisar  19.250030  72.859070
9                    Devipada  19.224690  72.866050
10                   Dombivli  19.212750  73.083240
11   Eastern Suburbs (Mumbai)  19.005347  72.855803
12                  Ghatkopar  19.086523  72.909008
13                   Goregaon  19.164550  72.849460
14                 Grant Road  18.959290  72.831080
15                 Jogeshwari  19.137920  72.849410
16                       Juhu  19.014920  72.845220
17                     Kalyan  18.953937  72.820367
18                  Kandivali  19.211982  72.837573
19                 Kanjurmarg  19.131380  72.935680
20                      Kausa  19.127580  72.825390
21                      Kurla  19.064980  72.880690
22  Mahavir Nagar (Kandivali)  19.211982  72.837573
23                   Mankhurd  19.048530  72.932220
24       Matharpacady, Mumbai  19.044920  72.867205
25                  Mira Road  19.280032  72.867932
26              Mogra Village  22.681810  77.467780
27                     Mulund  19.171830  72.955650
28                     Mumbra  19.190540  73.022660
29               Pestom sagar  19.070640  72.902170
30            Seven Bungalows  19.131342  72.816342
31                 Shil Phata  19.146580  73.040050
32               Sion, Mumbai  19.043590  72.864120
33           Sonapur, Bhandup  19.163940  72.935440
34             Thakur village  19.210200  72.875410
35       Tilak Nagar (Mumbai)  18.996160  72.852790
36                      Uttan  26.866340  80.938840
37                      Vashi  19.084650  72.904810
38                   Vikhroli  19.111090  72.927810
39                     Wadala  19.017200  72.858170
40   Western Suburbs (Mumbai)  19.197010  72.827680
41                      Worli  19.007440  72.816880

In [16]:
Ml_df.to_csv("Ml_df.csv", index=False)

<h1>Create a map of Mumbai with neighborhoods on top</h1>

In [17]:
address = 'Mumbai, India'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Mumbai, India {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Mumbai, India 18.9387711, 72.8353355.


In [18]:
map_Ml = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood in zip(Ml_df['Latitude'], Ml_df['Longitude'], Ml_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_Ml)  
    
map_Ml

In [19]:
map_Ml.save('map_Ml.html')

<h1>Using Foursquare to get neighborhood venues</h1>

In [20]:
CLIENT_ID = 'GYRPOICLY4F3THPAA2LIM3V4UVDC1KFZ35U3QDOR20VCPIUW' # your Foursquare ID
CLIENT_SECRET = 'KX1KY0DRXZWRAD1Y232QP11544JMFNE2IGOW4KT4TMRE5TCG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GYRPOICLY4F3THPAA2LIM3V4UVDC1KFZ35U3QDOR20VCPIUW
CLIENT_SECRET:KX1KY0DRXZWRAD1Y232QP11544JMFNE2IGOW4KT4TMRE5TCG


In [21]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(Ml_df['Latitude'], Ml_df['Longitude'], Ml_df['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [22]:
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(2670, 7)


Neighborhood   Latitude  Longitude                     VenueName  \
0      Andheri  19.118459  72.841763             Merwans Cake shop   
1      Andheri  19.118459  72.841763  Radha Krishna Veg Restaurant   
2      Andheri  19.118459  72.841763                      Naturals   
3      Andheri  19.118459  72.841763                   Temptations   
4      Andheri  19.118459  72.841763              Narayan Sandwich   

   VenueLatitude  VenueLongitude      VenueCategory  
0      19.119300       72.845418             Bakery  
1      19.115130       72.843060  Indian Restaurant  
2      19.111204       72.837255     Ice Cream Shop  
3      19.113767       72.841337     Ice Cream Shop  
4      19.121398       72.850270     Sandwich Place

In [23]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                               
Andheri                         100        100        100            100   
Anushakti Nagar                  16         16         16             16   
Baiganwadi                        8          8          8              8   
Bandra                          100        100        100            100   
Bhandup                          24         24         24             24   
Borivali                        100        100        100            100   
Charkop                          53         53         53             53   
Chembur                         100        100        100            100   
Dahisar                          66         66         66             66   
Devipada                         88         88         88             88   
Dombivli                         18         18         18             18   
Eastern Suburbs (Mumbai)         45         45         45             45   
Ghatkopar                        81         81         81             81   
Goregaon                        100        100        100            100   
Grant Road                      100        100        100            100   
Jogeshwari                       50         50         50             50   
Juhu                            100        100        100            100   
Kalyan                          100        100        100            100   
Kandivali                        75         75         75             75   
Kanjurmarg                       26         26         26             26   
Kausa                           100        100        100            100   
Kurla                            91         91         91             91   
Mahavir Nagar (Kandivali)        75         75         75             75   
Mankhurd                         13         13         13             13   
Matharpacady, Mumbai             50         50         50             50   
Mira Road                        36         36         36             36   
Mulund                           73         73         73             73   
Mumbra                            4          4          4              4   
Pestom sagar                     83         83         83             83   
Seven Bungalows                 100        100        100            100   
Shil Phata                        5          5          5              5   
Sion, Mumbai                     67         67         67             67   
Sonapur, Bhandup                 43         43         43             43   
Thakur village                   62         62         62             62   
Tilak Nagar (Mumbai)             21         21         21             21   
Uttan                            21         21         21             21   
Vashi                           100        100        100            100   
Vikhroli                        100        100        100            100   
Wadala                           97         97         97             97   
Western Suburbs (Mumbai)         79         79         79             79   
Worli                           100        100        100            100   

                           VenueLongitude  VenueCategory  
Neighborhood                                              
Andheri                               100            100  
Anushakti Nagar                        16             16  
Baiganwadi                              8              8  
Bandra                                100            100  
Bhandup                                24             24  
Borivali                              100            100  
Charkop                                53             53  
Chembur                               100            100  
Dahisar                                66             66  
Devipada                               88             88  
Dombivli                               18         

In [24]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 196 uniques categories.


<h4> Let's find out unique venues</h4>

In [25]:
venues_df['VenueCategory'].unique()[:50]

array(['Bakery', 'Indian Restaurant', 'Ice Cream Shop', 'Sandwich Place',
       'Fast Food Restaurant', 'Pub', 'Falafel Restaurant', 'Juice Bar',
       'Pizza Place', 'Coffee Shop', 'Multiplex', 'Seafood Restaurant',
       'Breakfast Spot', 'Chinese Restaurant', 'Snack Place',
       'Maharashtrian Restaurant', 'Café', 'American Restaurant',
       'Cocktail Bar', 'Bar', 'BBQ Joint', 'Diner', 'Lounge',
       'Asian Restaurant', 'Electronics Store', 'Spa', 'Department Store',
       "Women's Store", 'Residential Building (Apartment / Condo)',
       'Gym / Fitness Center', 'Smoke Shop', 'Food Truck',
       'Athletics & Sports', 'Vegetarian / Vegan Restaurant', 'Park',
       'Liquor Store', 'Fish Market', 'Martial Arts Dojo', 'Tea Room',
       'Dessert Shop', 'Food', 'Plaza', 'Bus Station', 'Sports Bar',
       'Platform', 'Food & Drink Shop', 'Music Venue', 'Tunnel',
       'Fried Chicken Joint', 'Gym'], dtype=object)

In [26]:
"Neighborhood" in venues_df['VenueCategory'].unique()

True

<h1>Analysing neighborhoods</h1>

In [27]:
# one hot encoding
Ml_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Ml_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Ml_onehot.columns[-1]] + list(Ml_onehot.columns[:-1])
Ml_onehot = Ml_onehot[fixed_columns]

print(Ml_onehot.shape)
Ml_onehot.head()

(2670, 197)


Neighborhoods  ATM  Afghan Restaurant  American Restaurant  Antique Shop  \
0       Andheri    0                  0                    0             0   
1       Andheri    0                  0                    0             0   
2       Andheri    0                  0                    0             0   
3       Andheri    0                  0                    0             0   
4       Andheri    0                  0                    0             0   

   Arcade  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0       0            0                    0                 0   
1       0            0                    0                 0   
2       0            0                    0                 0   
3       0            0                    0                 0   
4       0            0                    0                 0   

   Athletics & Sports  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0       1     0    0   
1                   0          0           0       0     0    0   
2                   0          0           0       0     0    0   
3                   0          0           0       0     0    0   
4                   0          0           0       0     0    0   

   Basketball Court  Beach  Bed & Breakfast  Beer Garden  Big Box Store  \
0                 0      0                0            0              0   
1                 0      0                0            0              0   
2                 0      0                0            0              0   
3                 0      0                0            0              0   
4                 0      0                0            0              0   

   Bike Rental / Bike Share  Bistro  Bookstore  Bowling Alley  \
0                         0       0          0              0   
1                         0       0          0              0   
2                         0       0          0              0   
3                         0       0          0              0   
4                         0       0          0              0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  Building  \
0                     0               0        0            0         0   
1                     0               0        0            0         0   
2                     0               0        0            0         0   
3                     0               0        0            0         0   
4                     0               0        0            0         0   

   Burger Joint  Burrito Place  Bus Station  Cafeteria  Café  Camera Store  \
0             0              0            0          0     0             0   
1             0              0            0          0     0             0   
2             0              0            0          0     0             0   
3             0              0            0          0     0             0   
4             0              0            0          0     0             0   

   Cheese Shop  Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  \
0            0                   0               0             0            0   
1            0                   0               0             0            0   
2            0                   0               0             0            0   
3            0                   0               0             0            0   
4            0                   0               0             0            0   

   College Auditorium  College Gym  Comedy Club  Comfort Food Restaurant  \
0                   0            0            0                        0   
1                   0            0            0                        0   
2                   0            0            0                        0   
3                   0            0            0                        0   
4                   0            0            0                        0   

   Concert Hall  Convenience Stor

In [28]:
Ml_grouped = Ml_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(Ml_grouped.shape)
Ml_grouped

(41, 197)


Neighborhoods  ATM  Afghan Restaurant  American Restaurant  \
0                     Andheri  0.0           0.000000             0.010000   
1             Anushakti Nagar  0.0           0.000000             0.000000   
2                  Baiganwadi  0.0           0.000000             0.000000   
3                      Bandra  0.0           0.000000             0.000000   
4                     Bhandup  0.0           0.000000             0.000000   
5                    Borivali  0.0           0.000000             0.000000   
6                     Charkop  0.0           0.000000             0.000000   
7                     Chembur  0.0           0.000000             0.000000   
8                     Dahisar  0.0           0.000000             0.000000   
9                    Devipada  0.0           0.000000             0.000000   
10                   Dombivli  0.0           0.000000             0.000000   
11   Eastern Suburbs (Mumbai)  0.0           0.000000             0.000000   
12                  Ghatkopar  0.0           0.012346             0.012346   
13                   Goregaon  0.0           0.000000             0.000000   
14                 Grant Road  0.0           0.000000             0.000000   
15                 Jogeshwari  0.0           0.000000             0.020000   
16                       Juhu  0.0           0.000000             0.000000   
17                     Kalyan  0.0           0.000000             0.000000   
18                  Kandivali  0.0           0.000000             0.000000   
19                 Kanjurmarg  0.0           0.000000             0.000000   
20                      Kausa  0.0           0.000000             0.020000   
21                      Kurla  0.0           0.000000             0.000000   
22  Mahavir Nagar (Kandivali)  0.0           0.000000             0.000000   
23                   Mankhurd  0.0           0.000000             0.000000   
24       Matharpacady, Mumbai  0.0           0.000000             0.000000   
25                  Mira Road  0.0           0.000000             0.000000   
26                     Mulund  0.0           0.000000             0.000000   
27                     Mumbra  0.0           0.000000             0.000000   
28               Pestom sagar  0.0           0.000000             0.000000   
29            Seven Bungalows  0.0           0.000000             0.010000   
30                 Shil Phata  0.2           0.000000             0.000000   
31               Sion, Mumbai  0.0           0.000000             0.000000   
32           Sonapur, Bhandup  0.0           0.000000             0.000000   
33             Thakur village  0.0           0.000000             0.000000   
34       Tilak Nagar (Mumbai)  0.0           0.000000             0.000000   
35                      Uttan  0.0           0.000000             0.000000   
36                      Vashi  0.0           0.000000             0.010000   
37                   Vikhroli  0.0           0.010000             0.010000   
38                     Wadala  0.0           0.000000             0.000000   
39   Western Suburbs (Mumbai)  0.0           0.000000             0.012658   
40                      Worli  0.0           0.000000             0.000000   

    Antique Shop    Arcade  Art Gallery  Arts & Crafts Store  \
0           0.00  0.000000          0.0             0.000000   
1           0.00  0.000000          0.0             0.000000   
2           0.00  0.000000          0.0             0.000000   
3           0.00  0.020000          0.0             0.000000   
4           0.00  0.041667          0.0             0.000000   
5           0.00  0.010000          0.0             0.000000   
6           0.00  0.000000          0.0             0.000000   
7           0.00  0.000000          0.0             0.000000   
8           0.00  0.015152          0.0             0.000000   
9           0.00  0.000000          0.0             0.000000   
10          0.00  0.055556         

<h4>Creating a data frame with shoping mall</h4>

In [29]:
Ml_mall = Ml_grouped[["Neighborhoods","Shopping Mall"]]

In [30]:
Ml_mall.head()

Neighborhoods  Shopping Mall
0          Andheri       0.000000
1  Anushakti Nagar       0.000000
2       Baiganwadi       0.000000
3           Bandra       0.000000
4          Bhandup       0.041667

<h1>Clustering Neighborhoods</h1>

In [31]:
# set number of clusters
kclusters = 3

Ml_clustering = Ml_mall.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Ml_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 1, 2, 0, 2, 0, 0], dtype=int32)

In [32]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
Ml_merged = Ml_mall.copy()

# add clustering labels
Ml_merged["Cluster Labels"] = kmeans.labels_

In [33]:
Ml_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
Ml_merged.head()

Neighborhood  Shopping Mall  Cluster Labels
0          Andheri       0.000000               0
1  Anushakti Nagar       0.000000               0
2       Baiganwadi       0.000000               0
3           Bandra       0.000000               0
4          Bhandup       0.041667               1

In [41]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Ml_merged = Ml_merged.join(Ml_df.set_index("Neighborhood"), on="Neighborhood")

print(Ml_merged.shape)
Ml_merged.head() # check the last columns!

(41, 5)


Neighborhood  Shopping Mall  Cluster Labels   Latitude  Longitude
0          Andheri       0.000000               0  19.118459  72.841763
1  Anushakti Nagar       0.000000               0  19.042830  72.927340
2       Baiganwadi       0.000000               0  19.062940  72.926630
3           Bandra       0.000000               0  19.054370  72.840170
4          Bhandup       0.041667               1  19.145560  72.948560

In [38]:
# sort the results by Cluster Labels
print(Ml_merged.shape)
Ml_merged.sort_values(["Cluster Labels"], inplace=True)
Ml_merged

(41, 5)


Neighborhood  Shopping Mall  Cluster Labels   Latitude  \
0                     Andheri       0.000000               0  19.118459   
34       Tilak Nagar (Mumbai)       0.000000               0  18.996160   
31               Sion, Mumbai       0.000000               0  19.043590   
30                 Shil Phata       0.000000               0  19.146580   
27                     Mumbra       0.000000               0  19.190540   
25                  Mira Road       0.000000               0  19.280032   
24       Matharpacady, Mumbai       0.000000               0  19.044920   
23                   Mankhurd       0.000000               0  19.048530   
22  Mahavir Nagar (Kandivali)       0.000000               0  19.211982   
21                      Kurla       0.000000               0  19.064980   
18                  Kandivali       0.000000               0  19.211982   
17                     Kalyan       0.000000               0  18.953937   
16                       Juhu       0.000000               0  19.014920   
15                 Jogeshwari       0.000000               0  19.137920   
20                      Kausa       0.000000               0  19.127580   
2                  Baiganwadi       0.000000               0  19.062940   
11   Eastern Suburbs (Mumbai)       0.000000               0  19.005347   
10                   Dombivli       0.000000               0  19.212750   
1             Anushakti Nagar       0.000000               0  19.042830   
14                 Grant Road       0.000000               0  18.959290   
9                    Devipada       0.000000               0  19.224690   
8                     Dahisar       0.000000               0  19.250030   
3                      Bandra       0.000000               0  19.054370   
6                     Charkop       0.000000               0  19.208660   
33             Thakur village       0.032258               1  19.210200   
26                     Mulund       0.041096               1  19.171830   
4                     Bhandup       0.041667               1  19.145560   
37                   Vikhroli       0.030000               1  19.111090   
19                 Kanjurmarg       0.038462               1  19.131380   
35                      Uttan       0.047619               1  26.866340   
36                      Vashi       0.030000               1  19.084650   
38                     Wadala       0.020619               2  19.017200   
13                   Goregaon       0.020000               2  19.164550   
29            Seven Bungalows       0.010000               2  19.131342   
28               Pestom sagar       0.012048               2  19.070640   
5                    Borivali       0.010000               2  19.229360   
7                     Chembur       0.010000               2  19.062180   
39   Western Suburbs (Mumbai)       0.012658               2  19.197010   
12                  Ghatkopar       0.012346               2  19.086523   
32           Sonapur, Bhandup       0.023256               2  19.163940   
40                      Worli       0.020000               2  19.007440   

    Longitude  
0   72.841763  
34  72.852790  
31  72.864120  
30  73.040050  
27  73.022660  
25  72.867932  
24  72.867205  
23  72.932220  
22  72.837573  
21  72.880690  
18  72.837573  
17  72.820367  
16  72.845220  
15  72.849410  
20  72.825390  
2   72.926630  
11  72.855803  
10  73.083240  
1   72.927340  
14  72.831080  
9   72.866050  
8   72.859070  
3   72.840170  
6   72.826120  
33  72.875410  
26  72.955650  
4   72.948560  
37  72.927810  
19  72.935680  
35  80.938840  
36  72.904810  
38  72.858170  
13  72.849460  
29  72.816342  
28  72.902170  
5   72.857510  
7   72.902410  
39  72.827680  
12  72.909008  
32  72.935440  
40  72.816880

In [42]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Ml_merged['Latitude'], Ml_merged['Longitude'], Ml_merged['Neighborhood'], Ml_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [43]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

<h1>Examining Clusters</h1>

<h4>Cluster 0</h4>

In [44]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 0]

Neighborhood  Shopping Mall  Cluster Labels   Latitude  \
0                     Andheri            0.0               0  19.118459   
1             Anushakti Nagar            0.0               0  19.042830   
2                  Baiganwadi            0.0               0  19.062940   
3                      Bandra            0.0               0  19.054370   
6                     Charkop            0.0               0  19.208660   
8                     Dahisar            0.0               0  19.250030   
9                    Devipada            0.0               0  19.224690   
10                   Dombivli            0.0               0  19.212750   
11   Eastern Suburbs (Mumbai)            0.0               0  19.005347   
14                 Grant Road            0.0               0  18.959290   
15                 Jogeshwari            0.0               0  19.137920   
16                       Juhu            0.0               0  19.014920   
17                     Kalyan            0.0               0  18.953937   
18                  Kandivali            0.0               0  19.211982   
20                      Kausa            0.0               0  19.127580   
21                      Kurla            0.0               0  19.064980   
22  Mahavir Nagar (Kandivali)            0.0               0  19.211982   
23                   Mankhurd            0.0               0  19.048530   
24       Matharpacady, Mumbai            0.0               0  19.044920   
25                  Mira Road            0.0               0  19.280032   
27                     Mumbra            0.0               0  19.190540   
30                 Shil Phata            0.0               0  19.146580   
31               Sion, Mumbai            0.0               0  19.043590   
34       Tilak Nagar (Mumbai)            0.0               0  18.996160   

    Longitude  
0   72.841763  
1   72.927340  
2   72.926630  
3   72.840170  
6   72.826120  
8   72.859070  
9   72.866050  
10  73.083240  
11  72.855803  
14  72.831080  
15  72.849410  
16  72.845220  
17  72.820367  
18  72.837573  
20  72.825390  
21  72.880690  
22  72.837573  
23  72.932220  
24  72.867205  
25  72.867932  
27  73.022660  
30  73.040050  
31  72.864120  
34  72.852790

<h4>Cluster 1</h1>

In [45]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 1]

Neighborhood  Shopping Mall  Cluster Labels  Latitude  Longitude
4          Bhandup       0.041667               1  19.14556   72.94856
19      Kanjurmarg       0.038462               1  19.13138   72.93568
26          Mulund       0.041096               1  19.17183   72.95565
33  Thakur village       0.032258               1  19.21020   72.87541
35           Uttan       0.047619               1  26.86634   80.93884
36           Vashi       0.030000               1  19.08465   72.90481
37        Vikhroli       0.030000               1  19.11109   72.92781

<h4>Cluster 2</h4>

In [47]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 2]

Neighborhood  Shopping Mall  Cluster Labels   Latitude  \
5                   Borivali       0.010000               2  19.229360   
7                    Chembur       0.010000               2  19.062180   
12                 Ghatkopar       0.012346               2  19.086523   
13                  Goregaon       0.020000               2  19.164550   
28              Pestom sagar       0.012048               2  19.070640   
29           Seven Bungalows       0.010000               2  19.131342   
32          Sonapur, Bhandup       0.023256               2  19.163940   
38                    Wadala       0.020619               2  19.017200   
39  Western Suburbs (Mumbai)       0.012658               2  19.197010   
40                     Worli       0.020000               2  19.007440   

    Longitude  
5   72.857510  
7   72.902410  
12  72.909008  
13  72.849460  
28  72.902170  
29  72.816342  
32  72.935440  
38  72.858170  
39  72.827680  
40  72.816880

<h4>Observations</h4>
<p>We can see cluster 0 has no shopping malls, cluster 1 has most of shoping malls and cluster 2 has moderate shopping malls. So. it would be better to start shopping mall in cluster 0.</p>
<p>But lets see other categories which would be potential ventures.</p>

In [48]:
venues_df['VenueCategory'].value_counts()

Indian Restaurant                           375
Café                                        144
Fast Food Restaurant                        140
Coffee Shop                                 113
Pizza Place                                 107
Ice Cream Shop                               91
Chinese Restaurant                           86
Restaurant                                   72
Bar                                          62
Lounge                                       55
Dessert Shop                                 54
Bakery                                       53
Pub                                          51
Snack Place                                  47
Vegetarian / Vegan Restaurant                47
Sandwich Place                               47
Train Station                                41
Gym / Fitness Center                         40
Asian Restaurant                             36
Italian Restaurant                           36
Seafood Restaurant                      

In [49]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = Ml_grouped['Neighborhoods']

for ind in np.arange(Ml_grouped.shape[0]):
    row_categories = Ml_grouped.iloc[ind, :].iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 1:] = row_categories_sorted.index.values[0:num_top_venues]

# neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted

(41, 6)


Neighborhoods 1st Most Common Venue 2nd Most Common Venue  \
0                     Andheri     Indian Restaurant           Snack Place   
1             Anushakti Nagar    Chinese Restaurant                  Park   
2                  Baiganwadi              Platform                   Gym   
3                      Bandra     Indian Restaurant                  Café   
4                     Bhandup     Indian Restaurant           Pizza Place   
5                    Borivali     Indian Restaurant                  Café   
6                     Charkop  Fast Food Restaurant           Coffee Shop   
7                     Chembur     Indian Restaurant  Fast Food Restaurant   
8                     Dahisar    Chinese Restaurant     Indian Restaurant   
9                    Devipada     Indian Restaurant        Ice Cream Shop   
10                   Dombivli                  Café           Pizza Place   
11   Eastern Suburbs (Mumbai)     Indian Restaurant           Coffee Shop   
12                  Ghatkopar     Indian Restaurant  Fast Food Restaurant   
13                   Goregaon  Fast Food Restaurant     Indian Restaurant   
14                 Grant Road     Indian Restaurant        Ice Cream Shop   
15                 Jogeshwari     Indian Restaurant        Ice Cream Shop   
16                       Juhu     Indian Restaurant  Fast Food Restaurant   
17                     Kalyan     Indian Restaurant            Restaurant   
18                  Kandivali     Indian Restaurant  Fast Food Restaurant   
19                 Kanjurmarg     Indian Restaurant         Train Station   
20                      Kausa                  Café     Indian Restaurant   
21                      Kurla     Indian Restaurant                   Bar   
22  Mahavir Nagar (Kandivali)     Indian Restaurant  Fast Food Restaurant   
23                   Mankhurd        Shop & Service     Food & Drink Shop   
24       Matharpacady, Mumbai     Indian Restaurant           Pizza Place   
25                  Mira Road     Indian Restaurant           Pizza Place   
26                     Mulund     Indian Restaurant  Fast Food Restaurant   
27                     Mumbra            Restaurant             Bookstore   
28               Pestom sagar     Indian Restaurant           Pizza Place   
29            Seven Bungalows                  Café                   Pub   
30                 Shil Phata                   ATM         Train Station   
31               Sion, Mumbai     Indian Restaurant                  Café   
32           Sonapur, Bhandup  Fast Food Restaurant     Indian Restaurant   
33             Thakur village     Indian Restaurant  Fast Food Restaurant   
34       Tilak Nagar (Mumbai)     Indian Restaurant                 Plaza   
35                      Uttan     Indian Restaurant                Bakery   
36                      Vashi     Indian Restaurant           Coffee Shop   
37                   Vikhroli     Indian Restaurant                  Café   
38                     Wadala     Indian Restaurant                  Café   
39   Western Suburbs (Mumbai)           Coffee Shop                   Pub   
40                      Worli     Indian Restaurant                  Café   

   3rd Most Common Venue          4th Most Common Venue 5th Most Common Venue  
0            Pizza Place           Gym / Fitness Center                   Bar  
1                    Pub                    Bus Station                  Food  
2                   Park                    Coffee Shop            Food Truck  
3                 Bakery                   Dessert Shop                   Bar  
4          Train Station                      Multiplex            Restaurant  
5         Ice Cream Shop                     Restaurant    Chinese Restaurant  
6                 Lounge                    Pizza Place     Indian Restaurant  
7            Pizza Place                           Café            Restaurant  
8   Fast Food Restaurant                 Ice Cream Shop      

In [50]:
neighborhoods_venues_sorted['1st Most Common Venue'].value_counts(normalize = True)

Indian Restaurant       0.682927
Fast Food Restaurant    0.073171
Café                    0.073171
Chinese Restaurant      0.048780
Restaurant              0.024390
Shop & Service          0.024390
Coffee Shop             0.024390
Platform                0.024390
ATM                     0.024390
Name: 1st Most Common Venue, dtype: float64

In [51]:
neighborhoods_venues_sorted['2nd Most Common Venue'].value_counts(normalize = True)

Fast Food Restaurant    0.170732
Café                    0.146341
Pizza Place             0.121951
Indian Restaurant       0.097561
Coffee Shop             0.073171
Ice Cream Shop          0.073171
Pub                     0.048780
Train Station           0.048780
Restaurant              0.024390
Bakery                  0.024390
Plaza                   0.024390
Bookstore               0.024390
Snack Place             0.024390
Park                    0.024390
Gym                     0.024390
Food & Drink Shop       0.024390
Bar                     0.024390
Name: 2nd Most Common Venue, dtype: float64

In [52]:
neighborhoods_venues_sorted['3rd Most Common Venue'].value_counts(normalize = True)

Fast Food Restaurant    0.170732
Pizza Place             0.146341
Café                    0.097561
Coffee Shop             0.097561
Ice Cream Shop          0.073171
Bakery                  0.048780
Snack Place             0.048780
Multiplex               0.048780
Pub                     0.048780
Lounge                  0.024390
Indian Restaurant       0.024390
Train Station           0.024390
Chinese Restaurant      0.024390
Seafood Restaurant      0.024390
Park                    0.024390
Gym / Fitness Center    0.024390
Clothing Store          0.024390
Art Gallery             0.024390
Name: 3rd Most Common Venue, dtype: float64

In [53]:
neighborhoods_venues_sorted['4th Most Common Venue'].value_counts(normalize = True)

Coffee Shop                      0.146341
Chinese Restaurant               0.121951
Café                             0.097561
Fast Food Restaurant             0.097561
Vegetarian / Vegan Restaurant    0.097561
Restaurant                       0.073171
Ice Cream Shop                   0.073171
Pizza Place                      0.048780
Dessert Shop                     0.024390
Diner                            0.024390
Italian Restaurant               0.024390
Brewery                          0.024390
Gym / Fitness Center             0.024390
Multiplex                        0.024390
Train Station                    0.024390
Bakery                           0.024390
Bus Station                      0.024390
Hookah Bar                       0.024390
Name: 4th Most Common Venue, dtype: float64

In [115]:
neighborhoods_venues_sorted['5th Most Common Venue'].value_counts(normalize = True)

Restaurant                       0.097561
Train Station                    0.097561
Coffee Shop                      0.097561
Ice Cream Shop                   0.073171
Pizza Place                      0.073171
Café                             0.048780
Pub                              0.048780
Vegetarian / Vegan Restaurant    0.048780
Dessert Shop                     0.048780
Clothing Store                   0.024390
Food Truck                       0.024390
Seafood Restaurant               0.024390
Indian Restaurant                0.024390
Bakery                           0.024390
Gym                              0.024390
Multiplex                        0.024390
Bar                              0.024390
Hotel                            0.024390
Tea Room                         0.024390
Factory                          0.024390
Sports Bar                       0.024390
Food & Drink Shop                0.024390
Diner                            0.024390
Italian Restaurant               0

<p>We can see Mumbai's most common venue is Indian Restuarant. We can select Indian Restuarant as other category but there are 375 Indian Restuarants in Mumbai. So, it won't be nice to start another Indian Restuarant because it would get huge competition from other Indian Restuarants. When we see Ice Cream Shop it is one of the top five most common venues. So, let's explore this category.</p>

In [116]:
len(Ml_grouped[Ml_grouped["Ice Cream Shop"] > 0])

31

In [54]:
Ml_ice = Ml_grouped[["Neighborhoods","Ice Cream Shop"]]

In [55]:
Ml_ice.head()

Neighborhoods  Ice Cream Shop
0          Andheri        0.040000
1  Anushakti Nagar        0.000000
2       Baiganwadi        0.000000
3           Bandra        0.020000
4          Bhandup        0.041667

In [56]:
# set number of clusters
kclusters = 3

kl_clustering = Ml_ice.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(kl_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 1, 1, 1, 0, 2, 0, 1, 0, 2], dtype=int32)

In [57]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
Ml_merged = Ml_ice.copy()

# add clustering labels
Ml_merged["Cluster Labels"] = kmeans.labels_

In [59]:
Ml_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
Ml_merged.head()

Neighborhood  Ice Cream Shop  Cluster Labels
0          Andheri        0.040000               0
1  Anushakti Nagar        0.000000               1
2       Baiganwadi        0.000000               1
3           Bandra        0.020000               1
4          Bhandup        0.041667               0

In [67]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Ml_merged = Ml_merged.join(Ml_df.set_index("Neighborhood"), on="Neighborhood")

print(Ml_merged.shape)
Ml_merged.head() # check the last columns!

ValueError: columns overlap but no suffix specified: Index(['Latitude', 'Longitude'], dtype='object')

In [61]:
# sort the results by Cluster Labels
print(Ml_merged.shape)
Ml_merged.sort_values(["Cluster Labels"], inplace=True)
Ml_merged

(41, 5)


Neighborhood  Ice Cream Shop  Cluster Labels   Latitude  \
0                     Andheri        0.040000               0  19.118459   
38                     Wadala        0.030928               0  19.017200   
35                      Uttan        0.047619               0  26.866340   
33             Thakur village        0.048387               0  19.210200   
32           Sonapur, Bhandup        0.046512               0  19.163940   
29            Seven Bungalows        0.030000               0  19.131342   
26                     Mulund        0.041096               0  19.171830   
25                  Mira Road        0.055556               0  19.280032   
22  Mahavir Nagar (Kandivali)        0.026667               0  19.211982   
39   Western Suburbs (Mumbai)        0.037975               0  19.197010   
18                  Kandivali        0.026667               0  19.211982   
17                     Kalyan        0.050000               0  18.953937   
16                       Juhu        0.050000               0  19.014920   
14                 Grant Road        0.060000               0  18.959290   
20                      Kausa        0.030000               0  19.127580   
12                  Ghatkopar        0.037037               0  19.086523   
8                     Dahisar        0.060606               0  19.250030   
4                     Bhandup        0.041667               0  19.145560   
6                     Charkop        0.037736               0  19.208660   
27                     Mumbra        0.000000               1  19.190540   
1             Anushakti Nagar        0.000000               1  19.042830   
37                   Vikhroli        0.010000               1  19.111090   
36                      Vashi        0.020000               1  19.084650   
2                  Baiganwadi        0.000000               1  19.062940   
34       Tilak Nagar (Mumbai)        0.000000               1  18.996160   
3                      Bandra        0.020000               1  19.054370   
31               Sion, Mumbai        0.014925               1  19.043590   
30                 Shil Phata        0.000000               1  19.146580   
40                      Worli        0.020000               1  19.007440   
13                   Goregaon        0.020000               1  19.164550   
7                     Chembur        0.010000               1  19.062180   
24       Matharpacady, Mumbai        0.000000               1  19.044920   
23                   Mankhurd        0.000000               1  19.048530   
21                      Kurla        0.000000               1  19.064980   
19                 Kanjurmarg        0.000000               1  19.131380   
11   Eastern Suburbs (Mumbai)        0.000000               1  19.005347   
28               Pestom sagar        0.024096               1  19.070640   
5                    Borivali        0.070000               2  19.229360   
9                    Devipada        0.079545               2  19.224690   
10                   Dombivli        0.111111               2  19.212750   
15                 Jogeshwari        0.100000               2  19.137920   

    Longitude  
0   72.841763  
38  72.858170  
35  80.938840  
33  72.875410  
32  72.935440  
29  72.816342  
26  72.955650  
25  72.867932  
22  72.837573  
39  72.827680  
18  72.837573  
17  72.820367  
16  72.845220  
14  72.831080  
20  72.825390  
12  72.909008  
8   72.859070  
4   72.948560  
6   72.826120  
27  73.022660  
1   72.927340  
37  72.927810  
36  72.904810  
2   72.926630  
34  72.852790  
3   72.840170  
31  72.864120  
30  73.040050  
40  72.816880  
13  72.849460  
7   72.902410  
24  72.867205  
23  72.932220  
21  72.880690  
19  72.935680  
11  72.855803  
28  72.902170  
5   72.857510  
9   72.866050  
10  73.083240  
15  72.849410

In [62]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Ml_merged['Latitude'], Ml_merged['Longitude'], Ml_merged['Neighborhood'], Ml_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [63]:
map_clusters.save('map_clusters.html')

<h4>Cluster </h4>

In [64]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 0]

Neighborhood  Ice Cream Shop  Cluster Labels   Latitude  \
0                     Andheri        0.040000               0  19.118459   
38                     Wadala        0.030928               0  19.017200   
35                      Uttan        0.047619               0  26.866340   
33             Thakur village        0.048387               0  19.210200   
32           Sonapur, Bhandup        0.046512               0  19.163940   
29            Seven Bungalows        0.030000               0  19.131342   
26                     Mulund        0.041096               0  19.171830   
25                  Mira Road        0.055556               0  19.280032   
22  Mahavir Nagar (Kandivali)        0.026667               0  19.211982   
39   Western Suburbs (Mumbai)        0.037975               0  19.197010   
18                  Kandivali        0.026667               0  19.211982   
17                     Kalyan        0.050000               0  18.953937   
16                       Juhu        0.050000               0  19.014920   
14                 Grant Road        0.060000               0  18.959290   
20                      Kausa        0.030000               0  19.127580   
12                  Ghatkopar        0.037037               0  19.086523   
8                     Dahisar        0.060606               0  19.250030   
4                     Bhandup        0.041667               0  19.145560   
6                     Charkop        0.037736               0  19.208660   

    Longitude  
0   72.841763  
38  72.858170  
35  80.938840  
33  72.875410  
32  72.935440  
29  72.816342  
26  72.955650  
25  72.867932  
22  72.837573  
39  72.827680  
18  72.837573  
17  72.820367  
16  72.845220  
14  72.831080  
20  72.825390  
12  72.909008  
8   72.859070  
4   72.948560  
6   72.826120

<h4>Cluster 1</h4>

In [65]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 1]

Neighborhood  Ice Cream Shop  Cluster Labels   Latitude  \
27                    Mumbra        0.000000               1  19.190540   
1            Anushakti Nagar        0.000000               1  19.042830   
37                  Vikhroli        0.010000               1  19.111090   
36                     Vashi        0.020000               1  19.084650   
2                 Baiganwadi        0.000000               1  19.062940   
34      Tilak Nagar (Mumbai)        0.000000               1  18.996160   
3                     Bandra        0.020000               1  19.054370   
31              Sion, Mumbai        0.014925               1  19.043590   
30                Shil Phata        0.000000               1  19.146580   
40                     Worli        0.020000               1  19.007440   
13                  Goregaon        0.020000               1  19.164550   
7                    Chembur        0.010000               1  19.062180   
24      Matharpacady, Mumbai        0.000000               1  19.044920   
23                  Mankhurd        0.000000               1  19.048530   
21                     Kurla        0.000000               1  19.064980   
19                Kanjurmarg        0.000000               1  19.131380   
11  Eastern Suburbs (Mumbai)        0.000000               1  19.005347   
28              Pestom sagar        0.024096               1  19.070640   

    Longitude  
27  73.022660  
1   72.927340  
37  72.927810  
36  72.904810  
2   72.926630  
34  72.852790  
3   72.840170  
31  72.864120  
30  73.040050  
40  72.816880  
13  72.849460  
7   72.902410  
24  72.867205  
23  72.932220  
21  72.880690  
19  72.935680  
11  72.855803  
28  72.902170

<h4>Cluster 2</h4>

In [66]:
Ml_merged.loc[Ml_merged['Cluster Labels'] == 2]

Neighborhood  Ice Cream Shop  Cluster Labels  Latitude  Longitude
5      Borivali        0.070000               2  19.22936   72.85751
9      Devipada        0.079545               2  19.22469   72.86605
10     Dombivli        0.111111               2  19.21275   73.08324
15   Jogeshwari        0.100000               2  19.13792   72.84941

<h1>Observations:</h1>
<p>Cluster 0 has moderate Ice Cream Shops, Cluster 1 has no Ice Cream Shops and Cluster 2 has most of Ice Cream Shops. So, it is better to start Ice Cream Shop in Cluster 1 and most of areas in Cluster 1 are good areas and they are with no Ice Cream Shops. So, it is better to start Ice Cream Shops in these cities. Ice Cream Shops are most visted places than Shopping Malls in Mumbai. So, it is better to start Ice Cream Shop than Shopping mall.</p>
